File: leetcode/solutions/0155.MinStack/MinStack.go
package solution

type MinStack struct {
	stack []element
}

type element struct {
	val int
	min int
}

func Constructor() MinStack {
	return MinStack{}
}

func (this *MinStack) Push(x int) {
	this.stack = append(this.stack, element{
		val: x,
		min: this.calcMin(x),
	})
}

func (this *MinStack) Pop() {
	if len(this.stack) > 0 {
		this.stack = this.stack[:len(this.stack)-1]
	}
}

func (this *MinStack) Top() int {
	if len(this.stack) > 0 {
		return this.stack[len(this.stack)-1].val
	}
	return 0
}

func (this *MinStack) GetMin() int {
	return this.stack[len(this.stack)-1].min
}

func (this *MinStack) calcMin(x int) int {
	if len(this.stack) == 0 {
		return x
	}

	currentMin := this.stack[len(this.stack)-1].min

	if currentMin > x {
		return x
	}

	return currentMin
}

File: leetcode/solutions/0155.MinStack/MinStack.ts
type element = {
  val: number;
  min: number;
};

export default class MinStack {
  private stack: element[] = [];

  push(x: number) {
    this.stack.push({
      min: this.calcMin(x),
      val: x,
    });
  }

  pop() {
    this.stack.pop();
  }

  top(): number {
    return this.stack[this.stack.length - 1].val;
  }

  getMin(): number {
    return this.stack[this.stack.length - 1]?.min;
  }

  calcMin(x: number): number {
    if (this.stack.length === 0) {
      return x;
    }

    return Math.min(this.stack[this.stack.length - 1].min, x);
  }
}
